### DISCUSSION TITLE
`npm run build` failed, struggling with it all day

### DISCUSSION THREAD

**Discussion by oran1248** [vercel/next.js#17052] [Help]
Hi,

Trying to build my app with `target: 'serverless'` in `next.config.js` (In-order to deploy on AWS Lambda).
When running `npm run build` I'm getting the following output:
```
Warning: Built-in CSS support is being disabled due to custom CSS configuration being detected.
See here for more info: https://err.sh/next.js/built-in-css-disabled  

info  - Using external babel configuration from C:\Users\User\Desktop\mysite\.babelrc
info  - Creating an optimized production build  
warn  - Compiled with warnings

./node_modules/next/dist/next-server/server/load-components.js
Critical dependency: the request of a dependency is an expression     

./node_modules/next/dist/next-server/server/load-components.js        
Critical dependency: the request of a dependency is an expression     

./node_modules/next/dist/next-server/server/require.js
Critical dependency: the request of a dependency is an expression     

./node_modules/next/dist/next-server/server/require.js
Critical dependency: the request of a dependency is an expression     

./node_modules/next/dist/next-server/server/require.js
Critical dependency: the request of a dependency is an expression     

./node_modules/argon2/argon2.js
Critical dependency: the request of a dependency is an expression     

./node_modules/require_optional/index.js
Critical dependency: the request of a dependency is an expression     

./node_modules/require_optional/index.js
Critical dependency: the request of a dependency is an expression     

./node_modules/require_optional/index.js
Critical dependency: the request of a dependency is an expression     

./node_modules/node-pre-gyp/lib/pre-binding.js
Critical dependency: the request of a dependency is an expression     

./node_modules/node-pre-gyp/lib/util/versioning.js
Critical dependency: the request of a dependency is an expression     

./node_modules/node-fetch/lib/index.js
Module not found: Can't resolve 'encoding' in 'C:\Users\User\Desktop\mysite\node_modules\node-fetch\lib'

./node_modules/mongodb/lib/operations/connect.js
Module not found: Can't resolve 'mongodb-client-encryption' in 'C:\Users\User\Desktop\mysite\node_modules\mongodb\lib\operations'

./node_modules/node-pre-gyp/lib/util/compile.js
Module not found: Can't resolve 'node-gyp' in 'C:\Users\User\Desktop\mysite\node_modules\node-pre-gyp\lib\util'

./node_modules/node-pre-gyp/lib/util/compile.js
Module not found: Can't resolve 'npm' in 'C:\Users\User\Desktop\mysite\node_modules\node-pre-gyp\lib\util'

./node_modules/node-pre-gyp/lib/util/nw-pre-gyp/index.html 1:0        
Module parse failed: Unexpected token (1:0)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
> <!doctype html>
| <html>
| <head>


> Build error occurred
Error: package.json does not exist at C:\package.json
    at Object.module.exports.wr/F.exports.find (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:188263:15)
    at Object.<anonymous> (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:154078:28)
    at Object.paWE (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:154141:30)
    at __webpack_require__ (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:23:31)
    at Object.xDbK (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:189175:64)
    at __webpack_require__ (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:23:31)
    at Object.gL7F (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:123042:67)
    at __webpack_require__ (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:23:31)
    at Module.IlR1 (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:53618:12)
    at __webpack_require__ (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:23:31)
    at Module.ScIc (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:81256:17)
    at __webpack_require__ (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:23:31)
    at C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:91:18
    at Object.<anonymous> (C:\Users\User\Desktop\mysite\.next\serverless\pages\_error.js:94:10)
    at Module._compile (internal/modules/cjs/loader.js:1133:30)       
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1153:10)
info  - Collecting page data .npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! client@0.1.0 build: `next build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the client@0.1.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\User\AppData\Roaming\npm-cache\_logs\2020-09-12T19_38_11_418Z-debug.log
```

What is the meaning of this error? 
```
Error: package.json does not exist at C:\package.json
```

And what about all the `Module not found: Can't resolve` errors?

Thanks!

**Comment by sambhav-gore** [comment]
do you have reproduction repository ?

**Reply by oran1248** [reply]
> do you have reproduction repository ?

Do you mean repo with all the code? No.

**Reply by oran1248** [reply]
@sambhav-gore so you have any idea?

**Reply by sambhav-gore** [reply]
@oran1248 if you can make a minimal reproduction repo with code I am sure someone can have a look at it and help you.

**Reply by oran1248** [reply]
@sambhav-gore  ok i will create one and update

**Reply by oran1248** [reply]
@sambhav-gore 
I've created a reproduction repo here: https://github.com/oran1248/reproduction-repo
please read README.md file, it explains the issue. thanks!

**Reply by sambhav-gore** [reply]
the root cause is because of this - https://github.com/oran1248/reproduction-repo/blob/06d1778abc41c0c38b8e25dd74d3631a6b7f4702/src/utils/security.js#L1 

if you remove that dependency, and for a moment if you change the function that uses that lib to directly return true, you can see the build will succeed. if you can use a JS-only solution to check encryption that might work as well.

unfortunately I am not good with Sagas so I am not really sure how the HOC in _app.js is configured. But the bottomline is that you can't use server-only dependencies in pages. Instead move them into an external api and call that using fetch from your Next.js app. OR make API routes and use it that way.

**Reply by oran1248** [reply]
> the root cause is because of this - https://github.com/oran1248/reproduction-repo/blob/06d1778abc41c0c38b8e25dd74d3631a6b7f4702/src/utils/security.js#L1
> 
> if you remove that dependency, and for a moment if you change the function that uses that lib to directly return true, you can see the build will succeed. if you can use a JS-only solution to check encryption that might work as well.
> 
> unfortunately I am not good with Sagas so I am not really sure how the HOC in _app.js is configured. But the bottomline is that you can't use server-only dependencies in pages. Instead move them into an external api and call that using fetch from your Next.js app. OR make API routes and use it that way.

I don't use a server-only dependencies in pages, the function that uses `argon2` is called only from server side code (i.e. `/api` folder).
Maybe this issue is related to another discussion I've opened, about mixing between server side and client side code in one file. 
here: https://github.com/vercel/next.js/discussions/17053

Also, how can you explain that an unused function is causing the build to fail? I really don't understand the process of building here i guess..

**Comment by oran1248** [comment]
@sambhav-gore I have updated the repo to contain a very simple reproduction code. No redux nor redux-saga.

**Reply by oran1248** [reply]
@sambhav-gore I don't know why, but splitting the file to tow seperated files -  one file importing `argon2` and one file importing `jsonwebtoken` solves the problem.
